### ISSUE TITLE
Auxiliary-Loss-Free Load Balancing 是怎么实现的

### ISSUE THREAD

**Issue by TerrenceVarada** [ScienceOne-AI/DeepSeek-671B-SFT-Guide#9]
During training, we keep monitoring the expert load on the whole batch of each training step. At the end of each step, we will decrease the bias term by 𝛾if its corresponding expert is overloaded, and increase it by 𝛾if its corresponding expert is underloaded, where 𝛾is a hyper-parameter called bias update speed. Through the dynamic adjustment, DeepSeek-V3 keeps balanced expert load during training, and achieves better performance than models that encourage load balance through pure auxiliary losses. 代码里好像没有显示的看到论文里的这部分内容。可以请教一下相关的实现思路吗。

**Comment by Opdoop** [comment]
You can see expert balance bias update implementation here:
> https://github.com/NVIDIA/Megatron-LM/blob/b428f80cd576f0e6a3b526c010c5b6014da69f7e/megatron/core/transformer/moe/moe_utils.py#L644-L661
```
def get_updated_expert_bias(tokens_per_expert, expert_bias, expert_bias_update_rate):
    """Update expert bias for biased expert routing. See https://arxiv.org/abs/2408.15664v1#

    Args:
        tokens_per_expert (torch.Tensor): The number of tokens assigned to each expert.
        expert_bias (torch.Tensor): The bias for each expert.
        expert_bias_udpate_rate (float): The update rate for the expert bias.
    """
    with torch.no_grad():
        # All Reduce Across TPxCPxDP group
        torch.distributed.all_reduce(
            tokens_per_expert,
            group=parallel_state.get_tensor_and_data_parallel_group(with_context_parallel=True),
        )
        average_tokens = tokens_per_expert.sum(dim=-1, keepdim=True) / tokens_per_expert.shape[-1]
        offset = average_tokens - tokens_per_expert
        updated_expert_bias = expert_bias + torch.sign(offset) * expert_bias_update_rate
        return updated_expert_bias
```

**Comment by zjc6666** [comment]
> You can see expert balance bias update implementation here:
> 
> > https://github.com/NVIDIA/Megatron-LM/blob/b428f80cd576f0e6a3b526c010c5b6014da69f7e/megatron/core/transformer/moe/moe_utils.py#L644-L661
> 
> ```
> def get_updated_expert_bias(tokens_per_expert, expert_bias, expert_bias_update_rate):
>     """Update expert bias for biased expert routing. See https://arxiv.org/abs/2408.15664v1#
> 
>     Args:
>         tokens_per_expert (torch.Tensor): The number of tokens assigned to each expert.
>         expert_bias (torch.Tensor): The bias for each expert.
>         expert_bias_udpate_rate (float): The update rate for the expert bias.
>     """
>     with torch.no_grad():
>         # All Reduce Across TPxCPxDP group
>         torch.distributed.all_reduce(
>             tokens_per_expert,
>             group=parallel_state.get_tensor_and_data_parallel_group(with_context_parallel=True),
>         )
>         average_tokens = tokens_per_expert.sum(dim=-1, keepdim=True) / tokens_per_expert.shape[-1]
>         offset = average_tokens - tokens_per_expert
>         updated_expert_bias = expert_bias + torch.sign(offset) * expert_bias_update_rate
>         return updated_expert_bias
> ```

请教下， 我查看deepseek v3模型中，每层gate中e_score_correction_bias参数的值在数值5附近， 如果按照这个上面这个实现逻辑的话， 我看最终我网络中的e_score_correction_bias最大值在0.5左右，你知道这是为什么吗？
还是说真是deepseek v3模型中，Auxiliary-Loss-Free这个方法，他们有一些trick存在呢？

我看代码中， bias是以下面的方式添加到scores中的， 原始的scores经过sigmoid函数之后， 他的值不应该在0-1之间吗， e_score_correction_bias 如果在5左右， 直觉上好像是一个很奇怪的事情

scores = logits.sigmoid()
scores_for_choice = scores.view(bsz * seq_len, -1) + self.e_score_correction_bias.unsqueeze(0)
